src/controller/tests/data_model/TestRead.cpp

```diff
@@ -991,6 +1007,256 @@ void TestReadInteraction::TestReadFabricScopedWithFabricFilter(nlTestSuite * apS
     NL_TEST_ASSERT(apSuite, ctx.GetExchangeManager().GetNumActiveExchanges() == 0);
 }
 
+namespace SubscriptionPathQuotaHelpers {
+class TestReadCallback : public app::ReadClient::Callback
+{
+public:
+    TestReadCallback() {}
+    void OnAttributeData(const app::ConcreteDataAttributePath & aPath, TLV::TLVReader * apData,
+                         const app::StatusIB & aStatus) override
+    {
+        if (apData != nullptr)
+        {
+            mAttributeCount++;
+        }
+    }
+
+    void OnDone() override { mOnDone++; }
+
+    void OnReportEnd() override { mOnReportEnd++; }
+
+    void OnError(CHIP_ERROR aError) override
+    {
+        mOnError++;
+        mLastError = aError;
+    }
+
+    void OnSubscriptionEstablished(uint64_t aSubscriptionId) override { mOnSubscriptionEstablishedCount++; }
+
+    void ClearCounters()
+    {
+        mAttributeCount                 = 0;
+        mOnReportEnd                    = 0;
+        mOnSubscriptionEstablishedCount = 0;
+        mOnDone                         = 0;
+        mOnError                        = 0;
+        mLastError                      = CHIP_NO_ERROR;
+    }
+
+    int32_t mAttributeCount                 = 0;
+    int32_t mOnReportEnd                    = 0;
+    int32_t mOnSubscriptionEstablishedCount = 0;
+    int32_t mOnDone                         = 0;
+    int32_t mOnError                        = 0;
+    CHIP_ERROR mLastError                   = CHIP_NO_ERROR;
+};
+
+void EstablishSubscriptions(nlTestSuite * apSuite, void * apContext, int32_t numSubs, int32_t pathPerSub,
+                            app::ReadClient::Callback * callback, std::vector<std::unique_ptr<app::ReadClient>> & readClients)
+{
+    TestContext & ctx  = *static_cast<TestContext *>(apContext);
+    auto sessionHandle = ctx.GetSessionBobToAlice();
+    std::vector<app::AttributePathParams> attributePaths(
+        pathPerSub, app::AttributePathParams(kTestEndpointId, TestCluster::Id, TestCluster::Attributes::Int16u::Id));
+
+    app::ReadPrepareParams readParams(sessionHandle);
+    readParams.mpAttributePathParamsList    = attributePaths.data();
+    readParams.mAttributePathParamsListSize = pathPerSub;
+    readParams.mMaxIntervalCeilingSeconds   = 1;
+    readParams.mKeepSubscriptions           = true;
+
+    for (int32_t i = 0; i < numSubs; i++)
+    {
+        std::unique_ptr<app::ReadClient> readClient =
+            std::make_unique<app::ReadClient>(app::InteractionModelEngine::GetInstance(), &ctx.GetExchangeManager(), *callback,
+                                              app::ReadClient::InteractionType::Subscribe);
+        NL_TEST_ASSERT(apSuite, readClient->SendRequest(readParams) == CHIP_NO_ERROR);
+        readClients.push_back(std::move(readClient));
+    }
+}
+
+} // namespace SubscriptionPathQuotaHelpers
+
+void TestReadInteraction::TestReadHandler_KillOverQuotaSubscriptions(nlTestSuite * apSuite, void * apContext)
```

Comment:
Please file an issue about adding tests for multi-fabric, since the current TestContext stuff doesn't support that.